Additional tests for presto

losipiuk opened the pull request:
Replacement for https://github.com/TeradataCenterForHadoop/presto-td/pull/13

martint commented:
I merged the commits for to_unixtime, color functions and null args to math functions. I'm reviewing the rest.

martint commented:
Would you mind rebasing this and removing the commits I've already cherry-picked?

losipiuk commented:
Working on this now. I will rebase to newest presto. And test current product tests set with new Dain's INSERT patch.

losipiuk commented:
I rebased onto origin/master and fixed some issues.
This patch compiles fine but 
for most tests to succeed it depends on following #2574, #3097 and #3092.

ilfrin commented:
@martint I believe this has been brought to a state that you asked for. Please move forward.

losipiuk commented:
@martint: I believe we addressed all your comments. Please reiterate when you have some time :)

martint commented on presto-product-tests-utils/src/test/java/com/facebook/presto/tests/querystats/TestHttpQueryStatsClient.java:
```diff
@@ -0,0 +1,115 @@
+import io.airlift.http.client.testing.TestingResponse;
+import io.airlift.json.ObjectMapperProvider;
+import org.assertj.core.api.Assertions;
+import org.mockito.ArgumentCaptor;
```
We generally avoid using mockito (or any other mocking framework) because it makes tests brittle and less readable. We leave it as a last resort for testing against libraries that don't have good abstraction (e.g., via interfaces, etc).

Take a look at airlift's TestingHttpClient, which we created for the specific purpose of testing code that relies on airlift's http client.

losipiuk replied:
Thanks.

I do not 100% agree that using mocking frameworks always makes code less readable and harder to maintain. It may be true in this case though. 
I replaced mockito with use of `TestingHttpClient`.

martint commented on pom.xml:
```diff
@@ -58,7 +58,7 @@
         <air.test.jvmsize>1792m</air.test.jvmsize>
 
-        <air.javadoc.lint>-missing</air.javadoc.lint>
+        <air.javadoc.lint>--missing</air.javadoc.lint>
```
This doesn't look right. This option is passed to the java compiler as an argument to the `-Xdoclint` setting. `--missing` is not a valid value.

losipiuk replied:
I do not know where did this change come from. Fixing.

martint commented on pom.xml:
```diff
@@ -278,6 +292,12 @@
             <dependency>
+                <groupId>org.codehaus.jackson</groupId>
+                <artifactId>jackson-mapper-asl</artifactId>
+                <version>1.9.13</version>
```
We use jackson 2.x throughout presto. Why do we need an older version?

losipiuk replied:
This does not seem to be used.
Removing.

martint commented on presto-product-tests-utils/pom.xml:
```diff
@@ -0,0 +1,79 @@
+            <exclusions>
+                <exclusion>
+                    <groupId>org.slf4j</groupId>
+                    <artifactId>slf4j-jdk14</artifactId>
```
Why is this being excluded? The presto codebase assumes that everything logs to java.util.logging. We need to be very careful with how logging is wired up, or it's easy to end up in dependency hell, missed logging, double logging, etc. Here's an explanation of how logging works in presto: https://groups.google.com/d/msg/presto-users/2hN1ZGw-SVA/jyyhf_H3SjcJ. Some details have changed since I wrote that, but the general idea stands.

losipiuk replied:
In this patch we used logback (through logback-classic.jar) as slf4j backend. I tried to replace it with java.util.logging backend but there is a problem with MDC. We store test name in MDC map and then use it in logging pattern. We could log thread id and and then do the matching by it, but it is less user friendly. So I would opt for keeping logback as logging backend here. This should not be a problem as this is not part of the product.

Api is the other thing. If you prefer I can change logging statements throughout presto-product-tests module from slf4j to java.util.logging.

Based on your diagram it looks like presto is using logback as logging backend. But it seems to me that it is not the case. Am I missing something?

martint replied:
> Based on your diagram it looks like presto is using logback as logging backend. But it seems to me that it is not the case. Am I missing something?

Yeah, hence my comment about some details having changed. We've since replaced it with a straight up java.util.Logging handler.

martint replied:
> We store test name in MDC map and then use it in logging pattern. We could log thread id and and then do the matching by it, but it is less user friendly. So I would opt for keeping logback as logging backend here. This should not be a problem as this is not part of the product.

Can you elaborate on how logging is used in the testing framework? 

In my experience, relying on thread-locals for carrying state (e.g., logback's MDC) is fragile -- once you make your code multi-threaded (executors, thread pools, etc), the model stops working gets much harder to deal with.

> Api is the other thing. If you prefer I can change logging statements throughout presto-product-tests module from slf4j to java.util.logging.

If we can manage to do that it'd be preferable.

losipiuk replied:
> >  We store test name in MDC map and then use it in logging pattern. We could log thread id and and then do the matching by it, but it is less user friendly. So I would opt for keeping logback as logging backend here. This should not be a problem as this is not part of the product.
Can you elaborate on how logging is used in the testing framework?

> In my experience, relying on thread-locals for carrying state (e.g., logback's MDC) is fragile -- once you make your code multi-threaded (executors, thread pools, etc), the model stops working gets much harder to deal with.

I agree that generally one should avoid thread-locals. Yet we decided that in case of tempto, despite the fact that limits possibility of using Executors in tests code, they do more good than harm. We use them for two things.

1. We use SLF4J MDC for storing executed test name. It is set and cleaned by `TestInitializationListener` which is TestNG listener coordinating test execution. By having testcases name in MDC we can run tests in parallel (which actually we do not do right now ;) ) and still know which log line is coming from which testcase.

2. Each running test has reference of its `TestContext` on thread local variable. With a bit of simplification `TestContext` is wrapper around guice's `Injector` holding all the external dependencies available to test. As you noted this is somewhat fragile, and requires test writers to follow specific guidelines, in terms of using threads in test-code (we provide `TestContextDsl` class for simplifying that). On the other hand it allows us to expose static helper functions like `QueryExecutor#query` and not pass actual dependencies everywhere explicitly. IMO This makes test writing easier. And tests more readable.

kokosing replied:
I removed usage of logback in favor of java util logging.

martint commented on presto-product-tests-utils/src/main/java/com/facebook/presto/tests/querystats/HttpQueryStatsClient.java:
```diff
@@ -0,0 +1,93 @@
+import static io.airlift.http.client.ResponseHandlerUtils.propagate;
+
+/**
+ * Implementation of {@link QueryStatsClient} based on Apache http client.
```
This comment is incorrect

losipiuk replied:
fixed

losipiuk replied:
fixed

martint commented on presto-product-tests-utils/src/main/java/com/facebook/presto/tests/querystats/HttpQueryStatsClient.java:
```diff
@@ -0,0 +1,93 @@
+        return httpClient.execute(request, new GetQueryStatsResponseHandler());
+    }
+
+    private final class GetQueryStatsResponseHandler
```
Use `FullJsonResponseHandler` instead of a custom handler. See how `StatementClient` does it.

losipiuk replied:
I think I cannot use it here. 
The reason for using custom handler here is that we do not parse whole response body (which is queryInfo) but only queryStats branch.

There is a problem deserializing QueryInfo as a whole. I am not an expert in jackson but we are getting error concerning `c.f.p.spi.type.Type`

```
com.fasterxml.jackson.databind.JsonMappingException: Can not construct instance of com.facebook.presto.spi.type.Type, problem: abstract types either need to be mapped to concrete types, have custom deserializer, or be instantiated with additional type information
 at [Source: [B@6d366c9b; line: 87, column: 9] (through reference chain: java.util.LinkedHashMap["a"])
````

martint commented on presto-product-tests/libs/download.sh:
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+set -e;
+
+HIVE_JDBC_DOWNLOAD_URL='http://bdch-ftp.td.teradata.com:8081/nexus/service/local/artifact/maven/redirect?r=test-framework&g=org.apache.hive&a=hive-jdbc-all&v=1.0-SNAPSHOT&e=jar'
```
Why do users need to download this dependency manually? It should be a maven dependency for this module

losipiuk replied:
Sure - this has to be addressed.
https://github.com/prestodb/presto-hive-apache/pull/17 is an approach to that :)

kokosing replied:
I removed that, now it requires https://github.com/prestodb/presto-hive-jdbc/pull/1 to be accepted and merged. Then presto-hive-jdbc needs to be deployed on maven repositories.

martint commented on presto-product-tests/README.md:
```diff
@@ -0,0 +1,37 @@
+
+## Test configuration
+
+Test assume that you have hadoop and presto cluster running. For setting up development environment we
```
Since this test suite requires that a presto cluster already exist, running it as part of the maven build seems like the wrong approach. I would expect this module to produce a binary artifact that bundles all the tests and that can be used to validate a presto instance.

I.e., the process would be:
- run the build, this produces the presto binary and test suite binary
- deploy the presto binary to a cluster
- run test suite binary against the cluster

sopel39 replied:
Could you describe your motivation for why you dislike launching product tests from Maven. Why is launching integration tests OK and product tests not? Currently we can launch product tests using both IDE (e.g IntelliJ) and Maven. Launching product tests from IntellJ is extremely useful when testing Presto in local development environment. But some people prefer to start tests from command line, therefore we would prefer to keep Maven way of launching product tests too.

However, as you mentioned, we could additionally make Maven to build an executable product tests JAR with all dependencies (that would probably also include Presto) that could be easily launched on test cluster. That would require a maven hack to produce such a jar but maybe that will be better than launching tests directly from maven in some cases.

martint replied:
> Could you describe your motivation for why you dislike launching product tests from Maven. Why is launching integration tests OK and product tests not? 

The main difference is that unit tests and integration tests in Presto test the *current* codebase, whereas the product tests in this suite are meant to test an external presto instance. So, while it's possible to run unit and integration tests as part of the build/release, it's not possible to do the same with product tests.

The way I see it, the main benefit from launching tests from the IDE is that you can step though the code to debug issues. This works well if it's unit or in-process integration tests. For tests that drive an external system, you'd only be debugging the testing framework or queries, no the target system. 

For us, requiring a source checkout to run the test suite is a big limitation. It's easier to package a tool and deploy it to our testing infrastructure than to push a copy of the code and tools necessary to be able to run a full maven build. It's also harder to know what version of the test suite is being exercised, etc.

> Currently we can launch product tests using both IDE (e.g IntelliJ) and Maven. Launching product tests from IntellJ is extremely useful when testing Presto in local development environment. But some people prefer to start tests from command line, therefore we would prefer to keep Maven way of launching product tests too.

Agreed, although this is orthogonal to whether it uses TestNG+surefire or a custom "application" that drives the test (like we do with the verifier).

kokosing replied:
We introduced tests runner which drives tests execution. It is called outside the maven.

martint commented:
Why two maven modules for this? Since the produce-test-utils are only meant to be used in product-tests, splitting them is premature. We can always do that later if necessary.

losipiuk commented:
> Why two maven modules for this? Since the produce-test-utils are only meant to be used in product-tests, splitting them is premature. We can always do that later if necessary.

This module is separate only because it contains test for `HttpQueryStatsClient`. Originally `HttpQueryStatsClient` was part of presto-product-tests but then test for it (`TestHttpQueryStatsClient`) then become a product-test. While it should be unit-test run during standard build.

martint commented:
> This module is separate only because it contains test for HttpQueryStatsClient. Originally HttpQueryStatsClient was part of presto-product-tests but then test for it (TestHttpQueryStatsClient) then become a product-test. While it should be unit-test run during standard build.

I see. I guess this is one of the downsides of using TestNG as the entry point of the product test suite.

kokosing commented:
Applying comments:
- removed the module product-tests-utils.
- rebased
- introduced usage of tests runner
- removed the usage of logback

Now this patch depends on https://github.com/prestodb/presto-hive-jdbc/pull/1 to be accepted, merged and released and on https://github.com/airlift/airbase to be released.

May ask you to review it further. Is there anything more to be fixed?

martint commented:
Awesome! The structure looks much better now. I just need to do one more pass on the code and we'll be good to go.

martint commented on pom.xml:
```diff
@@ -238,6 +240,18 @@
+
+            <dependency>
+                <groupId>com.facebook.presto</groupId>
+                <artifactId>presto-product-tests-utils</artifactId>
```
This module no longer exists

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/ProductTestFunctions.java:
```diff
@@ -0,0 +1,38 @@
+public class ProductTestFunctions
+        extends ProductTest
+{
+    @Test(groups = TestGroups.FUNCTIONS)
```
Static import FUNCTIONS?

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/SimpleQueryTest.java:
```diff
@@ -0,0 +1,88 @@
+                .contains(row(25));
+    }
+
+    @Test(groups = QUARANTINE)
```
Remove this one

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/SimpleQueryTest.java:
```diff
@@ -0,0 +1,88 @@
+        assertThat(1).isEqualTo(2);
+    }
+
+    @Test(enabled = false)
```
Remove this one

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/TemptoProductTestRunner.java:
```diff
@@ -0,0 +1,36 @@
+
+    private TemptoProductTestRunner()
+    {
+        throw new UnsupportedOperationException();
```
This is not needed. Just leave an empty private constructor

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/TestGroups.java:
```diff
@@ -0,0 +1,48 @@
+    public static final String CREATE_TABLE = "create_table";
+    public static final String CREATE_DROP_VIEW = "create_drop_view";
+    public static final String ALTER_TABLE = "alter_table";
+    public static final String INDIC_BUFFERS = "indic_buffers";
```
What's INDIC?

kokosing replied:
INDIC is an teradata internal format to store the data. I don't think it is needed to in presto/master, removing.

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/TestGroups.java:
```diff
@@ -0,0 +1,48 @@
+    public static final String NULL_CONNECTOR = "null";
+    public static final String SMOKE = "smoke";
+    public static final String JDBC = "jdbc";
+    public static final String QE = "qe";
```
What's QE?

kokosing replied:
query execution

kokosing replied:
Sorry, it stands for Query Engine.

martint replied:
Can you name `QUERY_ENGINE` for clarity, then?

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/TestGroups.java:
```diff
@@ -0,0 +1,48 @@
+    public static final String FUNCTIONS = "functions";
+    public static final String CLI = "cli";
+    public static final String HIVE_CONNECTOR = "hive_connector";
+    public static final String HIVE_CONNECTOR_014 = "hive_connector_014";
```
What's HIVE_CONNECTOR_014?

kokosing replied:
I don't know. Removed.

kokosing replied:
Clarification: I tried to investigate the reason why somebody added that group, but now it is not clear and current usage of it suggests it is not needed anymore.

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/functions/JsonFunctionsTests.java:
```diff
@@ -0,0 +1,64 @@
+public class JsonFunctionsTests
+        extends ProductTest
+{
+    @Test(groups = {JSON_FUNCTIONS})
```
`{}` are not needed

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestTablePartitioningInsertInto.java:
```diff
@@ -0,0 +1,121 @@
+
+    private static String partitionDataFile(int region)
+    {
+        return "com/facebook/presto/tests/hive/data/partitioned_nation/nation_region_" + region + ".textfile";
```
Where is this file read from? Is this relative to the classpath, to a specific directory, etc?

kokosing replied:
It is relative to classpath

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/HivePartitioningTest.java:
```diff
@@ -0,0 +1,45 @@
+    protected long getProcessedLinesCount(String sqlStatement, QueryResult queryResult)
+            throws SQLException
+    {
+        QueryStatsClient queryStatsClient = ThreadLocalTestContextHolder.testContext().getDependency(QueryStatsClient.class);
```
Using thread-locals to pass state around doesn't seem like a good idea. This usage is eerily similar to a static per-thread lookup, which is also not great because it suffers from most of the problems dependency injection tries to solve (hidden dependencies on external state, etc).

Ideally, you'd inject the `QueryStatsClient` into each test execution. You might want to consider modeling how state holders/dependencies are injected into tests after the way JMH does it via the `@Scope` annotation (not a pre-requisite for this PR, since it'd be a major change to Tempto)

kokosing replied:
I injected that into test instance with using annotation BeforeTestWithContext

martint commented on presto-product-tests/src/main/java/com/facebook/presto/tests/hive/TestTablePartitioningInsertInto.java:
```diff
@@ -0,0 +1,121 @@
+        implements RequirementsProvider
+{
+    @Inject
+    private MutableTablesState mutableTablesState;
```
We tend to avoid injectable private fields. We think of injection frameworks as a layer that helps assemble your objects, but it should be possible to do so manually if one wants to (e.g., in a unit test). If something is accessible to the injection framework, it should be accessible to regular code.

kokosing replied:
This state does need to be injected, it is already accessible statically.

kokosing commented:
@martint All the comments were applied and rebased.

kokosing commented:
Upgrade testng to 6.9.6 caused some unit tests to fail. I fixed that. Hope this time travis will pass.

martint commented:
Also, can you squash the commits as appropriate (i.e., get rid of the "work in progress" ones, squash the ones that address review comments into the corresponding commits, etc) before I do a final pass?

kokosing commented:
Rebased and squashed them all together.

kokosing commented:
I extracted https://github.com/facebook/presto/pull/3829 from this one. And now this patch depends on it.

martint commented:
#3829 is merged now. Can you fixed that last one comment I had (rename QE to QUERY_ENGINE)? I'll merge the PR after that.

kokosing commented:
Rebased and comment applied.

martint commented:
Merged, thanks!

kokosing commented:
Yay... :-) Thank you!

czw., 5.11.2015, 19:32 Martin Traverso użytkownik <notifications@github.com>
napisał:

> Merged, thanks!
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/facebook/presto/pull/3096#issuecomment-154146304>.
>
